### ISSUE TITLE
WinRM certificate implementation in HTTPs is very strange

### ISSUE THREAD

**Issue by 1RedOne** [PowerShell/PowerShell#2282]
## Steps to reproduce
- Use an internal CA to deploy a Cert template valid for WinRM
- Deploy the cert with an 8 hour validity period and renewal at 2 hours
- Deploy a script to auto-renew certs once every hour (b/c the default renewal period is 8 hours and ain't nobody got time for that)
- Enable WinRM w/ Https using `winrm qc -transport:https`
- Wait for cert to expire
## Expected behavior

We would expect to see one of two things happen here:
#### Cert Auto Renews and UI correctly reports this

or
## Actual behavior

In actuality, it appears that WinRM somehow does note that the certificate has been renewed, because it continues to accept WinRM connections over HTTPS with no issues, even after the certificate referenced under WSman\Listener has expired.  

However, everything in the UI reports outdated certificate information, for instance.

![On the left, the current valid cert, on the right, what PowerShell reports](http://i.imgur.com/9Sx8lpa.png)
_On the left, the current valid cert, on the right, what PowerShell reports_

These inconsistencies are all over the place.  WSman:\Service\CertificateThumbprint contains (in my case) the Thumbprint of the very first cert ever used with this machine, more than a week ago, also expired.  

The built in `winrm` command also seems to report the same cert, which was present when `winrm qc -transport:https` was first conducted.

![](http://i.imgur.com/7NKzaRT.png)
_Looking under wsman:\Service, a Cert thumbprint from a previous test is visible_

The strangest part of the whole thing is that even though WinRM References out of date and invalid certificate information all over, somehow only the correct and new cert is being presented when new Winrm connections come in (such as via `Enter-PSSession -UseSsl`.  
## Desired Change

The UI should correctly reflect the certificate thumbprint being used with WinRM, not refer to the original cert used to enable WinRM, as it does today.
## Environment data

``` powershell
> 
Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      5.0.10586.117                                                                           
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion                   10.0.10586.117                                                                          
CLRVersion                     4.0.30319.34011                                                                         
WSManStackVersion              3.0                                                                                     
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1                                                                                 




```

**Comment by mirichmo** [comment]
Investigating

**Comment by 1RedOne** [comment]
I wrote more about this on my blog!  👻👻👻 

[![](https://foxdeploy.files.wordpress.com/2016/09/the-case-of-the-ghost-certificate.png)](https://foxdeploy.com/2016/09/16/winrm-https-and-the-case-of-ghost-certificate)

https://foxdeploy.com/2016/09/16/winrm-https-and-the-case-of-ghost-certificate/

**Comment by mirichmo** [comment]
This issue appears to be with WinRM and not PowerShell. I notified the WinRM team and they will continue the investigation.

TFS:8956786

**Comment by 1RedOne** [comment]
For testing, I found that on a Server 2008r2 machine with WMF 2.0 that the same behavior exists.  WSMan lists old cert but correctly presents new cert. 

However in this configuration, the PsSession is dropped when the certificate changes.

**Comment by 1RedOne** [comment]
@mirichmo , could you ask the WinRm team to update this issue with their results?

**Comment by manojampalam** [comment]
I'll get back on this EOD 10/4

**Comment by manojampalam** [comment]
@1RedOne can you confirm that after the certificate renews, following shows the new bindings (though WinRM reports the stale thumbprint):
netsh http show sslcert ipport=0.0.0.0:5986

**Comment by hakabo** [comment]
My tests show the older certificate thumbprint.  (WinRM sessions over SSL confirmed still working using Enter-PSSession XX -UseSSL)

![capture](https://cloud.githubusercontent.com/assets/22237100/19105443/da300d14-8ad9-11e6-8383-913584ea638f.PNG)

**Comment by manojampalam** [comment]
@hakabo, thanks. I assume the older certificate is still valid. http.sys caches the certificate, so that's the reason the connection is working even when the cert is deleted from store. I expect the SSL endpoint to stop working after the cert has expired or after a machine reboot. 

Please let me know your observations.

**Comment by 1RedOne** [comment]
Hi guys,

 Just to confirm, I also see the expired certificate still listed when I run netsh as well.

Keep in mind that in my configuration, the cert will expire every eight hours, however, new HTTPS connections still work.  My hypothesis is that the cert is correctly updated, and that WinRM does present the correct cert, but that for some reason the WinRM PS drive and netsh commands list outdated information.

![image](https://cloud.githubusercontent.com/assets/1247626/19241228/8e351c3c-8edb-11e6-860e-b4cbaf86db9b.png)

**Comment by hakabo** [comment]
As its not explicitly been stated, the SSL connections continue to work after reboot.  

Screenshot below taken after a reboot. 

![capture](https://cloud.githubusercontent.com/assets/22237100/19345308/a8a70632-9135-11e6-9eb6-5b42e7f530b2.PNG)

Enter-PSSession -UseSSL can still establish connections remotely.

**Comment by 1RedOne** [comment]
I've tried capturing a trace using WireShark to help out, but

"Science Dog 'no clue what I'm doing'" .jpg

I've observed connections continuing to work no matter what, after reboot,
after restarting WinRM, etc.  It seems that somehow WinRM correctly grabs
the right certificate and uses this for connections, even if the UI always
displays the original cert used to configure WinRM.

**Comment by 1RedOne** [comment]
Hi guys, 

  I'm here for the MVP Summit this week.  Would love to chat with you guys if you have some time!

Thanks!

twitter.com/FoxDeploy

**Comment by jordanmills** [comment]
No resolution, just closed?

**Comment by SteveL-MSFT** [comment]
@jordanmills this issue is 2.5 years old without any new content.  If this is impacting you, I can reopen.

**Comment by 1RedOne** [comment]
The tldr is that windows certificate services served a role in authenticating the connection when a cert expired but a new one was valid for the devices involved.

Powershell should still be updated to udoate the cert thumbprint used, if it hasn't already been done. This would be a change for pwsh 6 though

**Comment by mrulke** [comment]
this is still an issue. all of the above is still correct.
this breaks connections from non-powershell utilities eg java.
just trying to curl the port gives you a connection reset from the server.
only way to get it working is to delete the winrm listener and then recreate it with the new thumbprint.

**Comment by jordanmills** [comment]
> 
@SteveL-MSFT SteveL-MSFT 
It looks like it's still impacting a lot of people.

**Comment by madburg** [comment]
Issue remains, @SteveL-MSFT SteveL-MSFT can you please re-open? 

Certificate lifecycle notifications where introduced back with Windows 8, Windows Server 2012, but from what I noticed they covered the system, services (e.g. online responder service auto renewing OCSP certs, cert stored under service account) and users certificates. 

However on an app/feature like IIS the IIS Admin remote management will not auto update the thumbprint of the renewed certificate and you can't connect once the cert that it is bound to expires, unfortunate but at least we know and expect this to be the case. However when it comes to WinRM there is a bug of some sort; either the lifecycle notifications does not encompass WinRM and/or WinRM source is doing something that is not documented and that is not a warm and fuzzy feeling. 

It continues to work, but no matter how you look up the certificate thumbprint WinRM has bound to it's Https Listener, it shows the thumbprint of a previously expired cert.... and as @1RedOne pointed out it will be the very first cert we ever configured to WinRM,

1) winrm get http://schemas.microsoft.com/wbem/wsman/1/config

2) (Get-ChildItem "WSMan:\localhost\Listener\$WinRMHttpsListener" | Where-Object Name -Like "CertificateThumbprint" | Select-Object -ExpandProperty Value) -replace " ",""

3) HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\WSMAN\Listener\*+HTTPS\certThumbprint

**Comment by mrulke** [comment]
@madburg, from what I can tell, Winrm has the thumbprint reference and only uses it to check the cert when updating. Otherwise, winrm knows very little about it if it is incorrect or missing, as the request gets dropped if anything happens.
It's also been a long time since I had to do this. Not for some reason; it just all works with  native Windows no matter what settings are in winrm

Also, when updating the cert, the Winrm wizard only updates if the old cert exists. The certification requirements are specific.
This is a snippet I had to create to update a bunch of VMM servers.


```
#Main Script
--
  | switch ($signed) {
  | 0{
  | $j = Invoke-Command  -Credential $cred -ComputerName $vmhosts.FQDN -Authentication Negotiate -AsJob -ScriptBlock {
  | # Gets the full host name from the current host
  | $fqdn = [System.Net.Dns]::GetHostByName($env:computerName).hostname
  | Write-Host -ForegroundColor Green "Connected to host $fqdn"
  | #Create New local Cert
  | Write-Host -ForegroundColor Green "Creating Self-Signed Cert"
  | New-SelfSignedCertificate -CertStoreLocation Cert:\LocalMachine\My -DnsName $fqdn -Subject $fqdn -NotAfter (Get-Date).AddYears(5) -ov cert
  | Write-Host -ForegroundColor Green "The cert thumbprint is $($cert.thumbprint)"
  | Write-Host -ForegroundColor Green "Removing old WinRM Listener"
  | winrm delete winrm/config/Listener?Address=*+Transport=HTTPS
  | #Create New WinRM Listener using New Cert
  | Write-Host -ForegroundColor Green "Creating New Listener with Thumbprint "$($cert.thumbprint)" "
  | winrm create winrm/config/Listener?Address=*+Transport=HTTPS "@{Hostname=""$fqdn""; CertificateThumbprint=""$($cert.thumbprint)""}" }
  | # Receive-Job -wait -name $j
  | }

```

**Comment by madburg** [comment]
Thank you @mrulke for sharing, the environment here is setup differently. 
On-prem ADCS implementation. All self-signed certs are deleted from Windows asset, not allowed. This is enforcement via a script executed by group policy. Any and all certificate requirement are handled by the on-prem PKI.

So the required certificate for WinRM are automatically on hosts, the WinRM client/server configuration is set by group policy,  which disables the _Http listener_ as well as FW rule only allows http**s** for WinRM) and that just leaves creating the initial WinRM 'Https' listener and bind it with the respective certificate; which is also done via group policy executing a a script which creates a scheduled task which runs the very same script (distinguishes what context it is being run in to know what steps to take; capable of self-healing, can update local code if any change is made centrally, and lastly write event(s) should any abnormality be detected) to ensure WinRM Https listener is setup asap provided the PKI certificate exist in the local cert store of the asset. We have 100% success ratio as certificates are enrolled at the near end of a build process and upon last reboot after build deployment completes the GPO's handle the rest instantly.

However the issue remains till this day whereby WinRM is not displaying a renewed certificate thumbprint yet remains functioning, however unknown in what security context? Everyone agrees that it should either break like the example I brought up with IIS Admin remote management, or WinRM should update the thumbprint if it is using the renewed certificate under the hood.

I cannot speak for @1RedOne if he had any self-signed certificates on the system he was testing with, but the Wireshark capture in his blog post clearly showed the client was presented with the updated/renewed PKI certificate not a self-sign one. We certainly do not have any self-sign certificates for WSMAN to fall back on to.

So maybe, not sure the feasibility of what I am about to say but stared at the details enough to believe there is something here.

Removing support of IPSPN support for kerberos, etc. keeping it to basics. We are domain joined, using FQDN hostnames to establish a remote WinRM connection to a host.

1) The initial connection from client to server is to establish authentication based on the allowed authentication types (kerberos, NTLM, etc..) WinHTTP established a TLS handshake nothing to do with whether we have WinRM Http**s** Listener configurated... Right? 

This handshake will exchange certificates to both parties provided each has an appropriate certificate to present to the other. So this is where I believe @1RedOne Wireshark capture displayed the client was presented with the servers renewed certificate (and the server presented the client certificate before hand just not extracted from the capture as it wasn't relevant to the blog post). So WinHTTP it is automatically selecting the latest available certificate that meets the "1.3.6.1.5.5.7.3.1 Server Authentication" requirement from the local cert store for SSL and presents it to the initiating client in the TLS handshake. It just so happens that certificate is the very same certificate he (in my case as well) uses for the WinRM Http**s** Listener.
Eventually the certificate expires and is the very same thumbprint that remains in the WinRM Http**s** Listener configuration. However the actually certificate is being renewed in the local cert store, which WinHTTP checks for the latest upon any connect. Right?

2) Quote#1 "Regardless of the transport protocol used (HTTP or HTTPS), WinRM **always** encrypts all PowerShell remoting communication after **initial authentication**." 
Quote#2 "Once initial authentication is complete, the WinRM encrypts the ongoing communication. When connecting over HTTPS, the TLS protocol is used to negotiate the encryption used to transport data. When connecting over HTTP, **message-level encryption** is determined by initial authentication protocol used."
https://learn.microsoft.com/en-us/powershell/scripting/security/remoting/winrm-security?view=powershell-7.4

The quotes above from MS is where it maybe useful that placement and additional verbiage is added for clarity.
WinRM encrypts the ongoing communication no matter what! So Certificate or NO Certificate we get message-level encryption  Right?

For "Regardless of the transport protocol used (HTTP or HTTPS)" I believe this is referencing to "WinRM Listener(s)" configuration for the transport protocol of HTTP and/or HTTPS (and/or as you can have both configured). I believe WinRm Listener(s) is exactly what they are referring to in that article because there is not configuration of WinHTTP (per se other than system wide supported encryption types, etc.. not a specific thumbprint).

3) Something I don't recall seeing in the past is a more detailed explanation of what the 'CertificateThumbprint' in the WinRM Listener configuration is for. We all believe it to be the Certificate to use and offer to establish a HTTPS tunnel from the very on set of the initial connection between client and server to ensure even 'Basic Authentication' wouldn't be in clear text. 

But looking at the MS WinRM article now (years later) and I quote "'CertificateThumbprint' - **Specifies the thumbprint of the service certificate.** This value represents a string of two-digit hexadecimal values found in the Thumbprint field of the certificate. This string contains the SHA-1 hash of the certificate. **Certificates are used in client certificate-based authentication. Certificates can be mapped only to local user accounts. They don't work with domain accounts.**" 

There are contradictory statements for ''CertificateThumbprint' Specifies the thumbprint of the service certificate. Is clearly what 
I remember being there, however the parts eluding to the certificate being used for client certificate-based auth. makes no sense as why would the WinRM SERVER configuration contain a client certificate-based cert for what would be mapped to a local account on the system... client certificate-based authentication certificates are passed on by clients to WinRM Listeners...
Clearly the case as when you move on to "WinRM **client** default configuration settings" the configuration has a setting named "Certificate" too "Allows the client to use client certificate-based authentication. Certificate-based authentication is a scheme in which the server authenticates a client identified by an X509 certificate. The default is True."

So my concern here is as followed, while it seems we are gifted by WinHTTP for selecting the latest qualifying server auth certificate thus we always successfully connect via TLS for our initial authentication and validating the endpoint is who we expect to be connecting too, We may not actually be establishing a SSL Tunnel between the hosts after initial authentication. Either we do have an SSL tunnel while the certificate is not expired and then it stops working after certificate expires, there just maybe no 'Event' log for the condition. 

"The credentials and the rest of the transmission are encrypted." This is refereeing to the message-level encryption which is determined by initial authentication protocol used.

"To add additional protection, you can configure the remote computer to use Secure Sockets Layer (SSL) instead of HTTP to listen for Windows Remote Management (WinRM) requests."
https://learn.microsoft.com/en-us/powershell/scripting/security/remoting/powershell-remoting-faq?view=powershell-7.4

Which is exactly what we are trying to achieve by binding a qualifying certificate to the 'WinRM Http**s** Listener'.

For clarity "When connecting over HTTP, message-level encryption is determined by initial authentication protocol used."

Basic authentication provides **_no encryption_.**
NTLM authentication uses an **RC4 cipher with a 128-bit key.** _not secure_
Kerberos authentication encryption is determined by the etype in the TGS ticket. This is AES-256 on modern systems. _OK depending on advisory_ 
CredSSP encryption is uses the TLS cipher suite that was negotiated in the handshake. _If TLS 1.3_ 

But with a SSL Tunnel we could encase the message-level encryption with up to TLS 1.3 with forward secrecy protocol and latest ciphers. Currently the asset certificates are dual role in that their used for endpoint verification as well as encryption. But that is changing where by we have rolled out a second ADCS implementation in order to properly rollout Ephemeral based certificates for endpoint services like SSL, VPN, WinRM, RDP as they benefit from the strength increase and the speed of modern cipher suites to keep them secure.

So I plead this issue get reopened and we get this figured out for the greater good. Thank you.
